csp/utils.py

```diff
@@ -6,85 +6,178 @@
-    return {
-        # Fetch Directives
-        'child-src': getattr(settings, 'CSP_CHILD_SRC', None),
-        'connect-src': getattr(settings, 'CSP_CONNECT_SRC', None),
-        'default-src': getattr(settings, 'CSP_DEFAULT_SRC', ["'self'"]),
-        'script-src': getattr(settings, 'CSP_SCRIPT_SRC', None),
-        'script-src-attr': getattr(settings, 'CSP_SCRIPT_SRC_ATTR', None),
-        'script-src-elem': getattr(settings, 'CSP_SCRIPT_SRC_ELEM', None),
-        'object-src': getattr(settings, 'CSP_OBJECT_SRC', None),
-        'style-src': getattr(settings, 'CSP_STYLE_SRC', None),
-        'style-src-attr': getattr(settings, 'CSP_STYLE_SRC_ATTR', None),
-        'style-src-elem': getattr(settings, 'CSP_STYLE_SRC_ELEM', None),
-        'font-src': getattr(settings, 'CSP_FONT_SRC', None),
-        'frame-src': getattr(settings, 'CSP_FRAME_SRC', None),
-        'img-src': getattr(settings, 'CSP_IMG_SRC', None),
-        'manifest-src': getattr(settings, 'CSP_MANIFEST_SRC', None),
-        'media-src': getattr(settings, 'CSP_MEDIA_SRC', None),
-        'prefetch-src': getattr(settings, 'CSP_PREFETCH_SRC', None),
-        'worker-src': getattr(settings, 'CSP_WORKER_SRC', None),
-        # Document Directives
-        'base-uri': getattr(settings, 'CSP_BASE_URI', None),
-        'plugin-types': getattr(settings, 'CSP_PLUGIN_TYPES', None),
-        'sandbox': getattr(settings, 'CSP_SANDBOX', None),
-        # Navigation Directives
-        'form-action': getattr(settings, 'CSP_FORM_ACTION', None),
-        'frame-ancestors': getattr(settings, 'CSP_FRAME_ANCESTORS', None),
-        'navigate-to': getattr(settings, 'CSP_NAVIGATE_TO', None),
-        # Reporting Directives
-        'report-uri': getattr(settings, 'CSP_REPORT_URI', None),
-        'report-to': getattr(settings, 'CSP_REPORT_TO', None),
-        'require-sri-for': getattr(settings, 'CSP_REQUIRE_SRI_FOR', None),
-        # Other Directives
-        'upgrade-insecure-requests': getattr(
-            settings, 'CSP_UPGRADE_INSECURE_REQUESTS', False),
-        'block-all-mixed-content': getattr(
-            settings, 'CSP_BLOCK_ALL_MIXED_CONTENT', False),
-    }
-
-
-def build_policy(config=None, update=None, replace=None, nonce=None):
-    """Builds the policy as a string from the settings."""
+HTTP_HEADERS = (
+    'Content-Security-Policy',
+    'Content-Security-Policy-Report-Only',
+)
+
+
+EXEMPTED_DEBUG_CODES = {
+    http_client.INTERNAL_SERVER_ERROR,
+    http_client.NOT_FOUND,
+}
 
+
+def from_settings():
+    policies = getattr(settings, 'CSP_POLICIES', defaults.POLICIES)
+    definitions = csp_definitions_update({}, defaults.POLICY_DEFINITIONS)
+    custom_definitions = getattr(
+        settings,
+        'CSP_POLICY_DEFINITIONS',
+        {'default': {}},
+    )
+     # Technically we're modifying Django settings here,
+     # but it shouldn't matter since the end result of either will be the same
+    deprecation._handle_legacy_settings(custom_definitions)
```

Comment:
Design Decision: Should we call _handle_legacy_settings based on whether the user has set CSP_POLICY_DEFINITIONS? i.e. only use the legacy settings if the new settings aren't used.

Reply:
I think this is probably best, but it require updating a whack-ton of tests, so maybe it could come in a separate PR.